psychopy/experiment/components/sound/__init__.py

```diff
@@ -250,38 +291,34 @@ def writeFrameCodeJS(self, buff):
         # because of the 'if' statement of the time test
         buff.setIndentLevel(-1, relative=True)
         buff.writeIndentedLines('}\n')
-        knownNote = (self.params['sound'] in knownNoteNames) or (self.params['sound'].val.isdigit())
-        if self.params['stopVal'].val in [None, 'None', '']:
-            if not knownNote:  # Known notes have no getDuration function because duration is infinite or not None
-                # infinite sounds
-                code = ('if (t >= (%(name)s.getDuration() + %(name)s.tStart) '
-                        '    && %(name)s.status === PsychoJS.Status.STARTED) {\n'
-                        '  %(name)s.stop();  // stop the sound (if longer than duration)\n'
-                        '  %(name)s.status = PsychoJS.Status.FINISHED;\n'
-                        '}\n')
-                buff.writeIndentedLines(code % self.params)
-        else:
-            # sounds with stop values
-            self.writeStopTestCodeJS(buff)
-            code = ("if (t >= %(name)s.tStart + 0.5) {\n"
-                    "  %(name)s.stop();  // stop the sound (if longer than duration)\n"
-                    "  %(name)s.status = PsychoJS.Status.FINISHED;\n"
-                    "}\n")
+        
+        # are we stopping this frame?
+        indented = self.writeStopTestCodeJS(buff, extra=" || %(name)s.isFinished")
+        if indented:
+            # stop playback
+            code = (
+                "// stop playback\n"
+            )
+            if self.params['syncScreenRefresh'].val:
+                code += (
+                    "psychoJS.window.callOnFlip(function(){ %(name)s.stop(); });\n"
```

Comment:
I don't think the end of the sound should be controlled by `syncScreenRefresh`, which is specifically for whether the start of the sound should sync to frame. When a sound is ending, it should end when it's supposed to end. See my PR to your branch.